File: tugas_3.java
import java.util.Scanner;

class Node {
    String data;
    Node next;
    Node prev;

    public Node(String data) {
        this.data = data;
    }
}

// Circular Doubly Linked List
class CircularDoublyLinkedList {
    private Node head;
    private int size;

    public CircularDoublyLinkedList() {
        head = null;
        size = 0;
    }

    // INSERT
    public void insert(String data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            newNode.next = newNode;
            newNode.prev = newNode;
        } else {
            Node tail = head.prev;

            tail.next = newNode;
            newNode.prev = tail;
            newNode.next = head;
            head.prev = newNode;
        }

        size++;
        System.out.println("Berita berhasil ditambahkan.\n");
    }

    // DELETE
    public void delete(int position) {
        if (head == null) {
            System.out.println("List kosong.\n");
            return;
        }

        if (position < 1 || position > size) {
            System.out.println("Nomor tidak valid.\n");
            return;
        }

        Node current = head;

        if (size == 1) {
            head = null;
        } else {
            for (int i = 1; i < position; i++) {
                current = current.next;
            }

            current.prev.next = current.next;
            current.next.prev = current.prev;

            if (position == 1) {
                head = current.next;
            }
        }

        size--;
        System.out.println("Berita berhasil dihapus.\n");
    }


    // DISPLAY LOOP FORWARD (pakai q)
    public void displayForwardLoop(Scanner input) {
        if (head == null) {
            System.out.println("Tidak ada berita.\n");
            return;
        }

        System.out.println("\n=== Teks Berjalan Forward ===");
        System.out.println("Ketik 'q' lalu Enter untuk keluar\n");

        Node current = head;
        final boolean[] stop = {false};

        Thread t = new Thread(() -> {
            while (true) {
                String s = input.nextLine();
                if (s.equalsIgnoreCase("q")) {
                    stop[0] = true;
                    break;
                }
            }
        });

        t.start();

        while (!stop[0]) {
            System.out.println(current.data);

            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {}

            current = current.next;
        }

        System.out.println("\nKembali ke menu...\n");
    }

    // DISPLAY LOOP BACKWARD (pakai q)
    public void displayBackwardLoop(Scanner input) {
        if (head == null) {
            System.out.println("Tidak ada berita.\n");
            return;
        }

        System.out.println("\n=== Teks Berjalan Backward ===");
        System.out.println("Ketik 'q' lalu Enter untuk keluar\n");

        Node current = head.prev;
        final boolean[] stop = {false};

        Thread t = new Thread(() -> {
            while (true) {
                String s = input.nextLine();
                if (s.equalsIgnoreCase("q")) {
                    stop[0] = true;
                    break;
                }
            }
        });

        t.start();

        while (!stop[0]) {
            System.out.println(current.data);

            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {}

            current = current.prev;
        }

        System.out.println("\nKembali ke menu...\n");
    }

    // DISPLAY AT
    public void displayAt(int position) {
        if (head == null) {
            System.out.println("List kosong.\n");
            return;
        }

        if (position < 1 || position > size) {
            System.out.println("Nomor tidak valid.\n");
            return;
        }

        Node current = head;

        for (int i = 1; i < position; i++) {
            current = current.next;
        }

        System.out.println("\nBerita ke-" + position + ":");
        System.out.println(current.data + "\n");
    }
}


// MAIN
public class tugas_3 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        CircularDoublyLinkedList cdll = new CircularDoublyLinkedList();

        while (true) {
            System.out.println("====== MENU TEKS BERJALAN ======");
            System.out.println("1. Insert berita");
            System.out.println("2. Hapus berita");
            System.out.println("3. Tampilkan forward loop");
            System.out.println("4. Tampilkan backward loop");
            System.out.println("5. Tampilkan berita tertentu");
            System.out.println("6. Exit");
            System.out.print("Pilih Opsi: ");

            int pilihan = input.nextInt();
            input.nextLine();

            switch (pilihan) {
                case 1:
                    System.out.print("Masukkan teks berita: ");
                    cdll.insert(input.nextLine());
                    break;

                case 2:
                    System.out.print("Masukkan nomor berita: ");
                    cdll.delete(input.nextInt());
                    break;

                case 3:
                    cdll.displayForwardLoop(input);
                    break;

                case 4:
                    cdll.displayBackwardLoop(input);
                    break;

                case 5:
                    System.out.print("Masukkan nomor berita: ");
                    cdll.displayAt(input.nextInt());
                    break;

                case 6:
                    System.out.println("Program selesai.");
                    input.close();
                    return;

                default:
                    System.out.println("Pilihan tidak valid.\n");
            }
        }
    }
}

File: tugas_3.py
import time

# Class Node
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


# Class Circular Doubly Linked List
class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    # Insert di akhir
    def insert(self, data):
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            new_node.next = new_node
            new_node.prev = new_node
        else:
            tail = self.head.prev

            tail.next = new_node
            new_node.prev = tail
            new_node.next = self.head
            self.head.prev = new_node

        self.size += 1
        print("Berita berhasil ditambahkan.\n")

    # Delete berdasarkan posisi
    def delete(self, position):
        if self.head is None:
            print("List kosong.\n")
            return

        if position < 1 or position > self.size:
            print("Nomor tidak valid.\n")
            return

        current = self.head

        if self.size == 1:
            self.head = None
        else:
            for _ in range(1, position):
                current = current.next

            current.prev.next = current.next
            current.next.prev = current.prev

            if position == 1:
                self.head = current.next

        self.size -= 1
        print("Berita berhasil dihapus.\n")

    # Running text forward (loop terus)
    def display_forward_loop(self):
        if self.head is None:
            print("Tidak ada berita.\n")
            return

        print("\n=== Running Text Forward ===")
        print("Tekan Ctrl + C untuk berhenti...\n")

        current = self.head

        try:
            while True:
                print(current.data)
                time.sleep(2)
                current = current.next
        except KeyboardInterrupt:
            print("\nRunning text dihentikan.\n")

    # Running text backward
    def display_backward_loop(self):
        if self.head is None:
            print("Tidak ada berita.\n")
            return

        print("\n=== Running Text Backward ===")
        print("Tekan Ctrl + C untuk berhenti...\n")

        current = self.head.prev

        try:
            while True:
                print(current.data)
                time.sleep(2)
                current = current.prev
        except KeyboardInterrupt:
            print("\nRunning text dihentikan.\n")

    # Tampilkan berita tertentu
    def display_at(self, position):
        if self.head is None:
            print("List kosong.\n")
            return

        if position < 1 or position > self.size:
            print("Nomor tidak valid.\n")
            return

        current = self.head

        for _ in range(1, position):
            current = current.next

        print(f"\nBerita ke-{position}:")
        print(current.data + "\n")


# MAIN PROGRAM
def main():
    cdll = CircularDoublyLinkedList()

    while True:
        print("====== MENU TEKS BERJALAN ======")
        print("1. Insert berita")
        print("2. Hapus berita")
        print("3. Tampilkan forward (loop)")
        print("4. Tampilkan backward (loop)")
        print("5. Tampilkan berita tertentu")
        print("6. Exit")

        pilihan = input("Pilih opsi: ")

        if pilihan == "1":
            berita = input("Masukkan teks berita: ")
            cdll.insert(berita)

        elif pilihan == "2":
            pos = int(input("Masukkan nomor berita: "))
            cdll.delete(pos)

        elif pilihan == "3":
            cdll.display_forward_loop()

        elif pilihan == "4":
            cdll.display_backward_loop()

        elif pilihan == "5":
            pos = int(input("Masukkan nomor berita: "))
            cdll.display_at(pos)

        elif pilihan == "6":
            print("Program selesai.")
            break

        else:
            print("Pilihan tidak valid.\n")


if __name__ == "__main__":
    main()